adam-core/src/main/scala/org/bdgenomics/adam/rdd/GenomicRDD.scala

```diff
@@ -744,41 +978,244 @@ trait GenomicRDD[T, U <: GenomicRDD[T, U]] {
     optPartitions: Option[Int] = None)(
       implicit tTag: ClassTag[T], xTag: ClassTag[X]): GenomicRDD[(Option[T], Iterable[X]), Z] = {
 
-    val (partitionSize, endSequences) = joinPartitionSizeAndSequences(optPartitions,
-      genomicRdd)
+    // did the user provide a set partition count?
+    // if no, take the max partition count from our rdds
+    val partitions = optPartitions.getOrElse(Seq(rdd.partitions.length,
+      genomicRdd.rdd.partitions.length).max)
+
+    val leftRdd = if (sorted) {
+      if (partitions != rdd.partitions.length) evenlyRepartition(partitions) else this
+    } else {
+      repartitionAndSort(partitions)
+    }
+
+    val preparedRdds = prepareForShuffleRegionJoin(leftRdd, genomicRdd)
+    val leftRddReadyToJoin = preparedRdds._1
+    val rightRddReadyToJoin = preparedRdds._2
+    assert(leftRddReadyToJoin.partitions.length == rightRddReadyToJoin.partitions.length)
+
+    // what sequences do we wind up with at the end?
+    val endSequences = sequences ++ genomicRdd.sequences
 
-    // key the RDDs and join
     GenericGenomicRDD[(Option[T], Iterable[X])](
       RightOuterShuffleRegionJoinAndGroupByLeft[T, X](endSequences,
-        partitionSize,
-        rdd.context).partitionAndJoin(flattenRddByRegions(),
-          genomicRdd.flattenRddByRegions()),
+        endSequences.records.map(_.length).sum / partitions, rdd.context)
+        .joinCoPartitionedRdds(leftRddReadyToJoin, rightRddReadyToJoin),
       endSequences,
       kv => {
         (kv._2.flatMap(v => genomicRdd.getReferenceRegions(v)) ++
           kv._1.toSeq.flatMap(v => getReferenceRegions(v))).toSeq
-      })
-      .asInstanceOf[GenomicRDD[(Option[T], Iterable[X]), Z]]
+      }, leftRdd.partitionMap).asInstanceOf[GenomicRDD[(Option[T], Iterable[X]), Z]]
+  }
+
+  /**
+   * The purpose of this method is to copartition two RDDs according to their ReferenceRegions.
```

Comment:
`The purpose of this method is to copartition two...` &rarr; `Copartition two...`